cli/src/commands/parse/pst.rs

```diff
@@ -0,0 +1,1275 @@
+    }
+}
+
+struct SubMessagesIter {
+    idx: usize,
+    total_number_of_messages: usize,
+    folder: PstFolder,
+}
+
+impl SubMessagesIter {
+    fn new(folder: PstFolder) -> Result<Self> {
+        let total_number_of_messages = folder.get_number_of_sub_messages()?;
+
+        Ok(Self {
+            idx: 0,
+            total_number_of_messages,
+            folder,
+        })
+    }
+}
+
+impl Iterator for SubMessagesIter {
+    type Item = Result<PstMessage>;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.idx < self.total_number_of_messages {
+            let item = self.folder.get_sub_message(self.idx);
+            self.idx += 1;
+            Some(item)
+        } else {
+            None
+        }
+    }
+}
+
+struct SubFoldersIter {
+    idx: usize,
+    total_number_of_folders: usize,
+    folder: PstFolder,
+}
+
+impl SubFoldersIter {
+    fn new(folder: PstFolder) -> Result<Self> {
+        let total_number_of_folders = folder.get_number_of_sub_folders()?;
+
+        Ok(Self {
+            idx: 0,
+            total_number_of_folders,
+            folder,
+        })
+    }
+}
+
+impl Iterator for SubFoldersIter {
+    type Item = Result<PstFolder>;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.idx < self.total_number_of_folders {
+            let item = self.folder.get_sub_folder(self.idx);
+            self.idx += 1;
+            Some(item)
+        } else {
+            None
+        }
+    }
+}
+
+struct AttachmentsIter {
+    idx: usize,
+    total_num_attachments: usize,
+    message: PstMessage,
+}
+
+impl Iterator for AttachmentsIter {
+    type Item = Result<PstAttachment>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.idx < self.total_num_attachments {
+            let next_attachment = self.message.get_attachment(self.idx);
+            self.idx += 1;
+            Some(next_attachment)
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(Clone)]
+struct PstMessage {
+    pub inner: *mut LibPffItem,
+    pub folder: FolderPath,
+}
+
+impl PstMessage {
+    // Useful for debugging
+    fn _get_message_id(&self) -> Result<String> {
+        let raw_headers = self
+            .get_transport_headers()?
+            .context("Could not read transport headers")?;
+        let (parsed_headers, _) = mailparse::parse_headers(raw_headers.as_bytes())?;
+        let message_id = Self::expect_header(&parsed_headers, "Message-ID")?;
+        Ok(message_id)
+    }
+
+    fn expect_header(headers: &[MailHeader], header: &str) -> Result<String> {
+        let value = headers
+            .get_first_value(header)
+            .map(|value| value.trim().to_owned())
+            .context(format!("{} header is missing", header))?;
+
+        if value.is_empty() {
+            Err(anyhow!("{} header is empty", header))
+        } else {
+            Ok(value)
+        }
+    }
+
+    pub fn into_new_email(self, mailbox: Mailbox) -> Result<NewEmail> {
```

Comment:
I think this method could be in a module completely by itself, as it should just use the public API of `PstMessage` and output `NewEmail`.

This would isolate all the business logic of converting a pst message to an email from the internals you've written.